components/ws-proxy/pkg/proxy/routes.go

```diff
@@ -118,46 +174,59 @@ func SupervisorIDEHostHandler(r *mux.Router, config *RouteHandlerConfig) {
 		})
 	})
 
-	targetResolver := func(cfg *Config, req *http.Request) (tgt *url.URL, err error) {
-		var dst url.URL
-		dst.Scheme = cfg.BlobServer.Scheme
-		dst.Host = cfg.BlobServer.Host
-		dst.Path = "/" + cfg.WorkspacePodConfig.SupervisorImage
-		return &dst, nil
-	}
-	r.NewRoute().Handler(proxyPass(config, targetResolver))
+	r.NewRoute().HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		redirectToBlobserve(w, req, config, config.Config.WorkspacePodConfig.SupervisorImage)
+	})
 }
 
 // TheiaRootHandler handles all requests under / that are not handled by any special case above (expected to be static resources only)
-func TheiaRootHandler(infoProvider WorkspaceInfoProvider) RouteHandler {
-	return func(r *mux.Router, config *RouteHandlerConfig) {
-		r.Use(logRouteHandlerHandler("TheiaRootHandler"))
-		var reslv targetResolver
-		if config.Config.BlobServer != nil {
-			reslv = dynamicTheiaResolver(infoProvider)
-		} else {
-			reslv = staticTheiaResolver
-		}
-		resolver := func(config *Config, req *http.Request) (*url.URL, error) {
-			return reslv(config, req)
-		}
+func TheiaRootHandler(r *mux.Router, config *RouteHandlerConfig, infoProvider WorkspaceInfoProvider) {
+	r.Use(logRouteHandlerHandler("TheiaRootHandler"))
+	r.Use(config.CorsHandler)
 
-		r.Use(config.CorsHandler)
-		r.NewRoute().HandlerFunc(
-			// Use the static theia server as primary source for resources
-			proxyPass(config, resolver,
-				// If the static theia server returns 404, re-route to the pod itself instead
-				withHTTPErrorHandler(
-					config.WorkspaceAuthHandler(
-						proxyPass(config, workspacePodResolver,
-							withWebsocketSupport(),
-							withOnProxyErrorRedirectToWorkspaceStartHandler(config.Config),
-						),
+	var reslv targetResolver
+	if config.Config.BlobServer != nil {
+		reslv = dynamicTheiaResolver(infoProvider)
+	} else {
+		reslv = staticTheiaResolver
+	}
+	resolver := func(config *Config, req *http.Request) (*url.URL, error) {
+		return reslv(config, req)
+	}
+	theiaProxyPass := // Use the static theia server as primary source for resources
+		proxyPass(config, resolver,
+			// If the static theia server returns 404, re-route to the pod itself instead
+			withHTTPErrorHandler(
+				config.WorkspaceAuthHandler(
+					proxyPass(config, workspacePodResolver,
+						withWebsocketSupport(),
+						withOnProxyErrorRedirectToWorkspaceStartHandler(config.Config),
 					),
 				),
 			),
 		)
+
+	if config.Config.BlobServer == nil {
+		r.NewRoute().HandlerFunc(theiaProxyPass)
+		return
 	}
+
+	r.NewRoute().HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		coords := getWorkspaceCoords(req)
+		info := infoProvider.WorkspaceInfo(coords.ID)
+		if info == nil {
+			theiaProxyPass.ServeHTTP(w, req)
+			return
+		}
+
+		resp, err := http.Get(fmt.Sprintf("%s://%s/%s%s", config.Config.BlobServer.Scheme, config.Config.BlobServer.Host, info.IDEImage, req.URL.Path))
```

Comment:
The default HTTP client doesn't have a timeout, hence may never return. It's be better if held a properly configured [http.Client](https://golang.org/pkg/net/http/#Client) instance in the proxy which we would use for requests like this.

`http.Client { Timeout: 30*time.Second }` will do.